crypto/crypto/src/multi_ed25519.rs

```diff
@@ -0,0 +1,547 @@
+        let payload_length = bytes.len();
+        // Special case for single keys to speed up deserialization.
+        if payload_length == ED25519_PRIVATE_KEY_LENGTH {
+            return Ed25519PrivateKey::try_from(bytes)
+                .map(|private_key| MultiEd25519PrivateKey(vec![private_key], 1u8));
+        }
+
+        let threshold = get_threshold(
+            payload_length,
+            ED25519_PRIVATE_KEY_LENGTH,
+            bytes[payload_length - 1],
+        )?;
+
+        let mut private_keys: Vec<Ed25519PrivateKey> = vec![];
+        let multi_key = bytes
+            .chunks(ED25519_PRIVATE_KEY_LENGTH)
+            .filter(|chunk| chunk.len() == ED25519_PRIVATE_KEY_LENGTH)
+            .try_for_each(|key_bytes| deserialize_and_push(key_bytes, &mut private_keys));
+
+        multi_key.map(|_| MultiEd25519PrivateKey(private_keys, threshold))
+    }
+}
+
+impl Length for MultiEd25519PrivateKey {
+    fn length(&self) -> usize {
+        self.to_bytes().len()
+    }
+}
+
+impl ValidKey for MultiEd25519PrivateKey {
+    fn to_bytes(&self) -> Vec<u8> {
+        self.to_bytes()
+    }
+}
+
+impl Genesis for MultiEd25519PrivateKey {
+    fn genesis() -> Self {
+        let mut buf = [0u8; ED25519_PRIVATE_KEY_LENGTH];
+        buf[ED25519_PRIVATE_KEY_LENGTH - 1] = 1;
+        MultiEd25519PrivateKey(vec![Ed25519PrivateKey::try_from(buf.as_ref()).unwrap()], 1)
+    }
+}
+
+//////////////////////
+// PublicKey Traits //
+//////////////////////
+
+/// Convenient method to create a MultiEd25519PublicKey from a single Ed25519PublicKey.
+impl From<&Ed25519PublicKey> for MultiEd25519PublicKey {
+    fn from(ed_public_key: &Ed25519PublicKey) -> Self {
+        MultiEd25519PublicKey(vec![ed_public_key.clone()], 1u8)
+    }
+}
+
+/// Implementing From<&PrivateKey<...>> allows to derive a public key in a more elegant fashion.
+impl From<&MultiEd25519PrivateKey> for MultiEd25519PublicKey {
+    fn from(private_key: &MultiEd25519PrivateKey) -> Self {
+        let public_key = private_key
+            .0
+            .iter()
+            .map(|ed_private_key| ed_private_key.public_key())
+            .collect();
+        MultiEd25519PublicKey(public_key, private_key.1)
+    }
+}
+
+/// We deduce PublicKey from this.
+impl PublicKey for MultiEd25519PublicKey {
+    type PrivateKeyMaterial = MultiEd25519PrivateKey;
+}
+
+/// Get the hash output of the serialized MultiEd25519PublicKey.
+impl std::hash::Hash for MultiEd25519PublicKey {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        let encoded_pubkey = self.to_bytes();
+        state.write(&encoded_pubkey);
+    }
+}
+
+// Those are required by the implementation of hash above.
+impl PartialEq for MultiEd25519PublicKey {
+    fn eq(&self, other: &MultiEd25519PublicKey) -> bool {
+        self.to_bytes() == other.to_bytes()
+    }
+}
+
+impl Eq for MultiEd25519PublicKey {}
+
+impl TryFrom<&[u8]> for MultiEd25519PublicKey {
```

Comment:
what are we implementing this for versus using serde + lcs?

Reply:
As with all algorithms, we do have `to_bytes` (which is fed to hashes) and need the opposite to also run sanity checks. We don't allow malicious keys to be generated and all `ValidKey`s assume that it was generated through the `TryFrom` trait. That's a general pattern in our crypto crate.

Reply:
pointer? my understanding was that we convert from object to bytes via lcs.

Reply:
@davidiw if signatures or keys can be deserialized by anything but through the `try_from` of their key/signature type, that's a [Critical](https://hackerone.com/libra) vulnerability.

Reply:
> pointer? my understanding was that we convert from object to bytes via lcs.

See https://github.com/libra/libra/blob/master/types/src/account_address.rs#L53